Runtime reachability of IPC endpoints should be defined within the .messages.in files

gavin-apple opened the pull request:
#### 35edbfdbf97e524b704fa8da7ef287848c19bae3
<pre>
Runtime reachability of IPC endpoints should be defined within the .messages.in files
<a href="https://bugs.webkit.org/show_bug.cgi?id=259552">https://bugs.webkit.org/show_bug.cgi?id=259552</a>
rdar://112970535

Reviewed by Alex Christensen.

We should define the runtime reachability of IPC endpoints consistently
within the .messages.in files. Previously, this was achieved by a series
of MESSAGE_CHECKS placed within the receiving handlers. This change
introduces an [EnabledIf=&quot;&quot;] attribute, which can be used within the
.messages.in files, which defines additional criteria which must be met at
runtime in order for the receiving handler to be invoked. Any messages received
which don&apos;t satisfy the additional criteria are dropped the same way as
if the endpoint had received a message it didn&apos;t know about.

* LayoutTests/TestExpectations:
* LayoutTests/ipc/restrictedendpoints/allow-access-attachmentElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-attachmentElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-modelElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-modelElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-webGPU-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-webGPU.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-attachmentElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-attachmentElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-modelElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-modelElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-updateQuotaBasedOnSpaceUsageForTesting-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webGPU-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webGPU.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webPush-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webPush.html:
* LayoutTests/ipc/restrictedendpoints/no-test-only-ipc-expected-crash-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/no-test-only-ipc-expected-crash.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::isWebGPUEnabled):
(WebKit::GPUConnectionToWebProcess::isWebGLEnabled):
(WebKit::GPUConnectionToWebProcess::updateWebGLEnabled):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updateWebGLEnabled):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::allowTestOnlyIPC):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(async_message_statement):
(sync_message_statement):
* Source/WebKit/Scripts/webkit/model.py:
(Message.__init__):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updateWebGLEnabled):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_limitsNavigationsToAppBoundDomains):
(WebKit::WebPageProxy::attachmentElementEnabled):
(WebKit::WebPageProxy::modelElementEnabled):
(WebKit::WebPageProxy::launchProcess):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updatePreferencesEnabledStateInGPUProcess):
(WebKit::WebProcessProxy::updateWebGPUEnabledStateInGPUProcess): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/267014@main">https://commits.webkit.org/267014@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b18b634e32d58706728f6df59dd3c73b34c8c49

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15333 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15636 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15998 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17089 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14393 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18153 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15736 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16991 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15941 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17823 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/15438 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13219 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13832 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20778 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13147 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13999 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17253 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14580 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14574 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12349 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15526 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13680 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3949 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3694 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18191 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15761 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14407 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3761 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Scripts/webkit/parser.py:
```diff
@@ -69,16 +69,22 @@ def parse(file):
 
+            runtime_enablement = None
+            if options_string:
+                match = re.search(r"(?:(?:, |^)+(?:EnabledIf='(.*)'))(?:, |$)?", options_string)
```
Add something in Source/WebKit/Scripts/webkit/test that covers this and run make in the same directory to update the test expectations.

gavin-apple committed:
Runtime reachability of IPC endpoints should be defined within the .messages.in files
https://bugs.webkit.org/show_bug.cgi?id=259552
rdar://112970535

Reviewed by Alex Christensen.

We should define the runtime reachability of IPC endpoints consistently
within the .messages.in files. Previously, this was achieved by a series
of MESSAGE_CHECKS placed within the receiving handlers. This change
introduces an [EnabledIf=""] attribute, which can be used within the
.messages.in files, which defines additional criteria which must be met at
runtime in order for the receiving handler to be invoked. Any messages received
which don't satisfy the additional criteria are dropped the same way as
if the endpoint had received a message it didn't know about.

* LayoutTests/TestExpectations:
* LayoutTests/ipc/restrictedendpoints/allow-access-attachmentElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-attachmentElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-modelElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-modelElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-webGPU-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/allow-access-webGPU.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-attachmentElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-attachmentElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-modelElement-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-modelElement.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-updateQuotaBasedOnSpaceUsageForTesting-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webGPU-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webGPU.html: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webPush-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/deny-access-webPush.html:
* LayoutTests/ipc/restrictedendpoints/no-test-only-ipc-expected-crash-expected.txt: Added.
* LayoutTests/ipc/restrictedendpoints/no-test-only-ipc-expected-crash.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::isWebGPUEnabled):
(WebKit::GPUConnectionToWebProcess::isWebGLEnabled):
(WebKit::GPUConnectionToWebProcess::updateWebGLEnabled):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updateWebGLEnabled):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::allowTestOnlyIPC):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(async_message_statement):
(sync_message_statement):
* Source/WebKit/Scripts/webkit/model.py:
(Message.__init__):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updateWebGLEnabled):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_limitsNavigationsToAppBoundDomains):
(WebKit::WebPageProxy::attachmentElementEnabled):
(WebKit::WebPageProxy::modelElementEnabled):
(WebKit::WebPageProxy::launchProcess):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updatePreferencesEnabledStateInGPUProcess):
(WebKit::WebProcessProxy::updateWebGPUEnabledStateInGPUProcess): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: https://commits.webkit.org/267014@main
